### PR TITLE
Balder/alternative dense tensor apply

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR

### CODE REVIEW

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp [hunk 1]
```diff
@@ -66,4 +71,24 @@ DenseTensorAddressCombiner::combineDimensions(const eval::ValueType &lhs,
...
+
+CommonDenseTensorCellsIterator::CommonDenseTensorCellsIterator(const Mapping & common,
+                                                               const Mapping & right,
+                                                               const eval::ValueType &type_in,
+                                                               CellsRef cells)
+    : _type(type_in),
+      _cells(cells),
+      _address(type_in.dimensions().size(), 0),
+      _common(common),
+      _mutable(right),
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp]: I suggest renaming _mutable to _right to preserve semantics across classes

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp [hunk 2]
```diff
@@ -66,4 +71,24 @@ DenseTensorAddressCombiner::combineDimensions(const eval::ValueType &lhs,
...
+      _accumulatedSize(_address.size())
+{
+    size_t multiplier = 1;
+    for (int32_t i(_address.size() - 1); i >= 0; i--) {
+        _accumulatedSize[i] = multiplier;
+        multiplier *= type_in.dimensions()[i].size;
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp]: please specify which of the types this is (I expect the rhs type)

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h [hunk 3]
```diff
@@ -71,4 +94,81 @@ class DenseTensorAddressCombiner
...
+                    cellIdx -= _accumulatedSize[rdim] * _type.dimensions()[rdim].size;
+                    rindex = 0; cindex = 0;
+                    curDimension--;
+                }
+            }
+        }
+    }
+    bool updateCommon(const Address & combined) {
+        for (const auto & m : _common) {
+            if (combined[m.first] >= _type.dimensions()[m.second].size) return false;
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h]: please use braces

File: eval/src/vespa/eval/tensor/dense/dense_tensor_apply.hpp [hunk 4]
```diff
@@ -10,21 +10,61 @@ namespace vespalib::tensor::dense {
...
+
+template <typename Function>
+std::unique_ptr<Tensor>
+apply(DenseTensorAddressCombiner & combiner, DirectDenseTensorBuilder & builder,
+      CommonDenseTensorCellsIterator & rhsIter, const DenseTensorView &lhs, Function &&func)
+{
+    for (DenseTensorCellsIterator lhsItr = lhs.cellsIterator(); lhsItr.valid(); lhsItr.next()) {
+        combiner.updateLeftAndCommon(lhsItr.address());
+        if (rhsIter.updateCommon(combiner.address())) {
+            rhsIter.for_each(combiner.address(), [&combiner, &func, &builder, &lhsItr](const DenseTensorCellsIterator::Address & combined, double rhsCell) {
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/dense/dense_tensor_apply.hpp]: does the lambda really need 'combiner'?

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h [hunk 5]
```diff
@@ -71,4 +94,81 @@ class DenseTensorAddressCombiner
...
+            cellIdx += address[i]*_accumulatedSize[i];
+        }
+        return cellIdx;
+    }
+public:
+    CommonDenseTensorCellsIterator(const Mapping & common, const Mapping & right,
+                                   const eval::ValueType &type_in, CellsRef cells);
+    ~CommonDenseTensorCellsIterator();
+    template <typename Func>
+    void for_each(Address & combined, Func && func) {
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h]: writing directly into the address of the combiner totally undermines the job of the combiner... Also, with 'combined' address being passed around, it gets hard to keep track of which ones are complete and which are incomplete.

**Review by havardpe — COMMENTED** [review]
The concept of the change seems reasonable.
I think the tension between the address combiner and the common dense tensor cells iterator makes it a bit complicated.

File: eval/src/vespa/eval/tensor/dense/dense_tensor_reduce.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        _sumDimSize = itr->size;
+        for (++itr; itr != oldType.dimensions().cend(); ++itr) {
+            _innerDimSize *= itr->size;
+        }
+    } else {
+        _outerDimSize = calcCellsSize(oldType);
+    }
+}
+
+}
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_reduce.cpp]: Addressed in #4587

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_store.cpp [hunk 7]
```diff
@@ -14,9 +14,9 @@ using vespalib::tensor::DenseTensor;
 using vespalib::tensor::DenseTensorView;
 using vespalib::tensor::MutableDenseTensorView;
 using vespalib::eval::ValueType;
+using std::make_unique;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/tensor/dense_tensor_store.cpp]: Addressed in #4587

File: eval/src/vespa/eval/tensor/dense/dense_tensor_view.cpp [hunk 8]
```diff
@@ -251,16 +251,26 @@ DenseTensorView::join(join_fun_t function, const Tensor &arg) const
...
 
 Tensor::UP
-DenseTensorView::reduce(join_fun_t op,
-                        const std::vector<vespalib::string> &dimensions) const
+DenseTensorView::reduce(join_fun_t op, const std::vector<vespalib::string> &dimensions) const
 {
-    return dense::reduce(*this,
-                         (dimensions.empty() ? _typeRef.dimension_names() : dimensions),
-                         op);
+    const std::vector<vespalib::string> & dims = (dimensions.empty() ? _typeRef.dimension_names() : dimensions);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_view.cpp]: Addressed in #4587

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp [hunk 9]
```diff
@@ -17,17 +17,22 @@ DenseTensorAddressCombiner::DenseTensorAddressCombiner(const eval::ValueType &lh
     auto rhsItrEnd = rhs.dimensions().cend();
     for (const auto &lhsDim : lhs.dimensions()) {
         while ((rhsItr != rhsItrEnd) && (rhsItr->name < lhsDim.name)) {
+            _right.emplace_back(_ops.size(), rhsItr-rhs.dimensions().cbegin());
             _ops.push_back(AddressOp::RHS);
             ++rhsItr;
         }
         if ((rhsItr != rhsItrEnd) && (rhsItr->name == lhsDim.name)) {
+            _left.emplace_back(_ops.size(), _left.size());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp]: Fixed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp [hunk 10]
```diff
@@ -17,17 +17,22 @@ DenseTensorAddressCombiner::DenseTensorAddressCombiner(const eval::ValueType &lh
...
+            _commonRight.emplace_back(_ops.size(), rhsItr-rhs.dimensions().cbegin());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp]: Fixed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp [hunk 11]
```diff
@@ -17,17 +17,22 @@ DenseTensorAddressCombiner::DenseTensorAddressCombiner(const eval::ValueType &lh
     auto rhsItrEnd = rhs.dimensions().cend();
     for (const auto &lhsDim : lhs.dimensions()) {
         while ((rhsItr != rhsItrEnd) && (rhsItr->name < lhsDim.name)) {
+            _right.emplace_back(_ops.size(), rhsItr-rhs.dimensions().cbegin());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp]: Fixed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp (hunk 1 above)
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.cpp]: Fixed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h [hunk 12]
```diff
@@ -19,34 +19,57 @@ class DenseTensorAddressCombiner
...
+    const Mapping & getCommonRight() const { return _commonRight; }
+    const Mapping & getRight() const { return _right; }
+
+    bool hasAnyRightOnlyDimensions() const { return ! _right.empty(); }
 
     const Address &address() const { return _combinedAddress; }
+    Address &address() { return _combinedAddress; }
 
-    bool combine(const CellsIterator &lhsItr, const CellsIterator &rhsItr) {
+    bool combine(const Address & lhs, const Address & rhs) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h]: Fixed -> Code removed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h (hunk 3 above)
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h]: Fixed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_apply.hpp (hunk 4 above)
Comment (baldersheim) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_apply.hpp]: No, gone

**Comment by baldersheim** [comment]
Supercedeed by #4594

**Pull request closed without merging**